### PR TITLE
Feature/organizations

### PR DESCRIPTION
**Pull request by eloigrndl**
Organizations class & builder

### CODE REVIEW

File: app/src/main/java/com/ncnf/organization/Organization.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    private List<String> adminIds;
+    private List<String>  events;
+
+    public Organization(String name, GeoPoint location, String address, String phoneNumber, String originalOwner) {
+
+        this(UUID.randomUUID(), name, location, address, phoneNumber, new ArrayList<String>() {{ add(originalOwner); }}, new ArrayList<>());
+    }
+
+    public Organization(UUID uuid, String name, GeoPoint location, String address, String phoneNumber, List<String> adminIds, List<String> events) {
+        if(isArrayEmpty(adminIds) || !checkAdminIds(adminIds)){
```
Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/organization/Organization.java:45]: I guess those conditions are an illustration of the  "fail safe default", as I suppose those checks will be performed before adding an Organization to the DB. I suppose we will always add an Organization first to the DB with a single admin, namely the creator

File: app/src/main/java/com/ncnf/Utils.java [hunk 2]
```diff
@@ -28,7 +28,7 @@
     /**
      * Keys related to events
      */
-    public static final String EVENTs_COLLECTION_KEY ="events/";
+    public static final String EVENTS_COLLECTION_KEY ="events/";
```
Comment (khcr) [inline app/src/main/java/com/ncnf/Utils.java:31]: I will sleep better better now

File: app/src/main/java/com/ncnf/organization/Organization.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        this.name = name;
+        this.location = location;
+        this.address = address;
+        this.phoneNumber = phoneNumber;
+        this.adminIds = adminIds;
+        this.events = events;
+    }
+
+    public boolean checkAdminIds(List<String> l){
+        for(String s : l){
```
Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/organization/Organization.java:59]: if you want `return l.stream().noneMatch(InputValidator::isStringEmpty);` is a bit cleaner

File: app/src/main/java/com/ncnf/organization/Organization.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+    private final UUID uuid;
+    private String name;
+    private GeoPoint location;
+    private String address;
+    private String phoneNumber;
+    private List<String> adminIds;
+    private List<String>  events;
+
+    public Organization(String name, GeoPoint location, String address, String phoneNumber, String originalOwner) {
```
Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/organization/Organization.java:39]: If you want to be consistent, you should either `@NonNull originalOwner` or check explicitly with `isEmptyString`, which I would add should be renamed to "isValidString" as you also check the nullity of the reference and not only its size

File: app/src/main/java/com/ncnf/organization/Organization.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        for(String s : l){
+            if(isStringEmpty(s)){
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public CompletableFuture<DatabaseResponse> saveToDB(DatabaseService db){
+        if(uuid == null || isArrayEmpty(adminIds)){
```
Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/organization/Organization.java:68]: The first check should be done in the constructor, the second is useless as it has already been performed in both constructors

File: app/src/main/java/com/ncnf/organization/Organization.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    public boolean checkAdminIds(List<String> l){
+        for(String s : l){
+            if(isStringEmpty(s)){
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public CompletableFuture<DatabaseResponse> saveToDB(DatabaseService db){
```
Comment (khcr) [inline app/src/main/java/com/ncnf/organization/Organization.java:67]: We usually pass the DatabaseService in the class constructor (e.g Event, User). Do we wish to be consistent with that ?

File: app/src/main/java/com/ncnf/organization/Organization.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+    public void setAddress(String address) {
+        this.address = address;
+    }
+
+    public void setPhoneNumber(String phoneNumber) {
+        this.phoneNumber = phoneNumber;
+    }
+
+    public void setAdminIds(List<String> adminIds) {
```
Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/organization/Organization.java:127]: If you do not want to destroy and recreate a List, you can do clearAll() followedby addAll

Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/organization/Organization.java:127]: You also might wanna check the input like you did prior

File: app/src/main/java/com/ncnf/organization/Organization.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        this.adminIds = new ArrayList<>(adminIds);
+    }
+
+    public void setEvents(List<String> events) {
```
Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/organization/Organization.java:131]: Same comments here. I would suggest creating a method `clearEvents` if you want to have an empty list of events for better clarity

File: app/src/main/java/com/ncnf/organization/Organization.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        this.events = new ArrayList<>(events);
+    }
+
+    public boolean addAdmin(String adminId){
+        if(isStringEmpty(adminId)){
+            return false;
+        }
+        return this.adminIds.add(adminId);
```
Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/organization/Organization.java:139]: I dont remember if ArrayLists allow duplicates, but if it is the case you should check if already exists before adding a new admin

File: app/src/main/java/com/ncnf/utilities/InputValidator.java [hunk 10]
```diff
@@ -80,4 +82,8 @@ public static boolean isValidPassword(String password){
     public static boolean isStringEmpty(String s){
         return s == null || s.length() == 0;
     }
+
+    public static boolean isArrayEmpty(List l){
```
Comment (HyPn0TiiK) [inline app/src/main/java/com/ncnf/utilities/InputValidator.java:86]: As mentioned before, maybe rename those two methods

File: app/src/main/java/com/ncnf/organization/OrganizationBuilder.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    public OrganizationBuilder(){
+        this.db = new DatabaseService();
+    }
+
+    public OrganizationBuilder(DatabaseService db){
+        if(db == null){
+            throw new IllegalArgumentException("Database is null");
+        }
+        this.db = db;
+    }
```
Comment (khcr) [inline app/src/main/java/com/ncnf/organization/OrganizationBuilder.java:39]: Maybe the class would be simpler to use if the method were static. We can save the burden to instantiate an object in order to build another one.
For example: (new Organization(myDb)).loadFromDB(uuid) // Organization.loadFromDB(myDb, data)

**Review by HyPn0TiiK — APPROVED** [review]
Looks like a complete and practical interface with the Organizer, good job

**Review by khcr — APPROVED** [review]

File: app/src/main/java/com/ncnf/organization/Organization.java (hunk 9 above)
Comment (khcr) [reply app/src/main/java/com/ncnf/organization/Organization.java:139]: It does. Maybe we should use a Set instead.

**Pull request merged**